File: lib/errors/scrape_exception.dart
class ScrapeException implements Exception {
  const ScrapeException(this.message, {this.source, this.cause});

  final String message;
  final String? source;
  final Object? cause;

  @override
  String toString() {
    final prefix = source == null
        ? 'ScrapeException'
        : 'ScrapeException[$source]';
    return '$prefix: $message';
  }
}

class CaptchaRequiredException extends ScrapeException {
  const CaptchaRequiredException(super.message, {super.source, super.cause});
}

class ScrapeTimeoutException extends ScrapeException {
  const ScrapeTimeoutException(super.message, {super.source, super.cause});
}

File: app/src/main/java/io/github/rustnomicon/sourceinstaller/errors/ScrapeException.kt
package io.github.rustnomicon.sourceinstaller.errors

open class ScrapeException(
    message: String,
    val source: String? = null,
    cause: Throwable? = null,
) : Exception(message, cause) {

    override fun toString(): String {
        val prefix = if (source == null) "ScrapeException" else "ScrapeException[$source]"
        return "$prefix: $message"
    }
}

class CaptchaRequiredException(
    message: String,
    source: String? = null,
    cause: Throwable? = null,
) : ScrapeException(message, source, cause)

class ScrapeTimeoutException(
    message: String,
    source: String? = null,
    cause: Throwable? = null,
) : ScrapeException(message, source, cause)
